aitelemetry/telemetrywrapper.go

```diff
@@ -220,6 +221,56 @@ func (th *telemetryHandle) TrackLog(report Report) {
 	th.client.Track(trace)
 }
 
+// TrackEvent function sends events to appinsights resource. It overrides a few of the existing columns
+// with app information.
+func (th *telemetryHandle) TrackEvent(aiEvent AiEvent) {
+
+	// Initialize new event message
+	event := appinsights.NewEventTelemetry(aiEvent.EventName)
+
+	// OperationId => resourceID (e.g.: NCID)
+	event.Tags.Operation().SetId(aiEvent.ResourceID)
+
+	// Copy the properties, if supplied
+	if aiEvent.Properties != nil {
+		for key, value := range aiEvent.Properties {
+			event.Properties[key] = value
+		}
+	}
+
+	// Acquire read lock to read metadata
+	th.rwmutex.RLock()
+	metadata := th.metadata
+
+	// Add metadata
+	if metadata.SubscriptionID != "" {
+		event.Tags.User().SetAccountId(metadata.SubscriptionID)
+
+		// AnonId => VMName
+		event.Tags.User().SetId(metadata.VMName)
+
+		// SessionId => VMID
+		event.Tags.Session().SetId(metadata.VMID)
+		event.Tags.Operation().SetParentId(th.appVersion)
```

Comment:
Can other non-metadata related fields be moved outside the lock? ( similar to the ones you have on line 268 )

Reply:
Moved non-metadata fields outside of the lock.